### DISCUSSION TITLE
Scheduled Transactions - Grant automatic execute

### DISCUSSION THREAD

**Discussion by DevAnyKR** [moneymanagerex/moneymanagerex#6568] [Q&A]
Even after the Recurring Transaction name has been changed to the Scheduled Transaction name, there are still questions.

![image](https://github.com/moneymanagerex/moneymanagerex/assets/110871727/10151331-2192-4399-b89a-b9de083058f7)

1. Checked Grant automatic execute
However, it does not register automatically.
That's why I'm copying it manually so far.

How to use automatic execute?

2. If set it to 6106, the remaining days will be shown as normal, but if  set it to 6107 and later(~9999), it will look like a 1491033 day delay. 🤣
this problem was said last year, but it still hasn't been fixed. just because no one actually uses it like that...

**Comment by n-stein** [comment]
> How to use automatic execute?

1. Set the Due Date and recurrence on the left
2. Set the Date Paid to the actual transaction date on the right. Fill the other mandatory fields.
3. Check "Grant Automatic Execute"

Note that Automatic Execute only runs when you first open MMEX, so you have to close and reopen MMEX to trigger it.

Automatic Execute checks for cases where Date Due <= today and enters the transaction, incrementing both Due Date and Date Paid by the recurrence. It will enter as many transactions as necessary until Date Paid > today.

**Reply by DevAnyKR** [reply]
previously registered recurring transaction has never worked so far.

Test iterative transaction I registered yesterday will work from today, but there is a serious problem.

It will be registered for infinite repetition as shown below.

![image](https://github.com/moneymanagerex/moneymanagerex/assets/110871727/17eb4554-7bcc-4d61-a159-9288ab6c733a)

So this part is something I didn't understand or I can't agree with.

1. registration time for the repeat transaction is 09:12:00. but, now it's 08:55:xx. It still have time, but a repeat transaction has been registered.

2. Today is March 8th. However, transactions after that will continue to be registered in DB.  😵

One more thing, booking transactions are very slow to operate. It takes 2-4 seconds to register one as shown. Repetitive input process performance is suspected.

3. Move home each time a repeat transaction is registered. this is unnecessary action.

4. Program cannot be terminated while a repeat transaction is being registered.
Unable to stop the transaction entry operation.
MMEX process must terminate by OS. 

So please let me know how to stop the infinite repeat registration.
I don't want to use the backup file because I registered quite a few deals yesterday. 🤣

5. And please let me know how I can handle the previously registered reservation transaction (it doesn't work).
recurring transaction settings were registered from 1.6.x and remained the same for a long time.

![image](https://github.com/moneymanagerex/moneymanagerex/assets/110871727/f5fce43a-d510-4d25-98af-c548a23d2f17)
payment date is March 22 because MMEX didn't register automatically, so I manually copied and entered the transaction. So far, I've done that.

😢

**Reply by DevAnyKR** [reply]
I modified the DB to stop repeat transactions. This is something regular users can't do. Exception handling is required to prevent malfunction caused by incorrect use.

I can finally see the screen I couldn't see because of item 3 above.

![image](https://github.com/moneymanagerex/moneymanagerex/assets/110871727/fa137f38-8bdc-4a07-bddb-cea12ed639dd)

🤣🤣🤣

**Reply by n-stein** [reply]
> It will be registered for infinite repetition as shown below

Yes, there is an overflow happening with dates > 6107-04-01 where the days until next occurrence wraps around and instead of being 4000 years in the future becomes 4000 years in the past so the system attempts to bring you current by logging the 4000 years of transactions it thinks are overdue. Seems to be occurring because wxTimeSpan::GetDays returns an int. Using GetSeconds returns a long long which doesn't overflow at 9999-12-31, so I will fix this.

> registration time for the repeat transaction is 09:12:00. but, now it's 08:55:xx. It still have time, but a repeat transaction has been registered.

Currently the Due Date is checked against wxDate::Today to see if a transaction is overdue. If Time is enabled we should check against wxDateTime::Now. If a user wants to disregard the time they can just set it to 00:00:00.

> booking transactions are very slow to operate. It takes 2-4 seconds to register one as shown
> Move home each time a repeat transaction is registered. this is unnecessary action.

Yes, I have noticed there is an unnecessary delay as well. We can reduce it by waiting until the end to refresh the home page.

**Reply by DevAnyKR** [reply]
> Yes, there is some sort of overflow happening with dates > 6107-04-01 where the days until next occurrence wraps around and instead of being 4000 years in the future becomes 4000 years in the past so the system attempts to bring you current by logging the 4000 years of transactions it thinks are overdue. Need to investigate this further.

Oh, it was 6107 when I asked, but I last revised it to 2100 yesterday.

> Yes, I have noticed there is an unnecessary delay as well. We can reduce it by waiting until the end to refresh the home page.

If the current screen is not the homepage, I don't think there is a need to go to the homepage and refresh it.
It is efficient to reload with the event handler only when the user moves home.

**Reply by n-stein** [reply]
> If the current screen is not the homepage, I don't think there is a need to go to the homepage and refresh it.
> It is efficient to reload with the event handler only when the user moves home.

Yes, I should have said refresh the current panel data.

**Reply by n-stein** [reply]
> And please let me know how I can handle the previously registered reservation transaction (it doesn't work).\nrecurring transaction settings were registered from 1.6.x and remained the same for a long time.\n\npayment date is March 22 because MMEX didn't register automatically, so I manually copied and entered the transaction. So far, I've done that.

Your Due Date is in 2050, MMEX will only automatically execute if the due date is <= today. That field is not when you want the automatic transactions to stop, it's when you want the next transaction to be entered.

You probably want to set Due Date = Date Paid (i.e. you want to enter the transaction in the register on its TRANSDATE) unless you want to log the transaction in the future. For example, if you always pay a bill on the 15th of the month and you want the transaction to be entered a week ahead of time to make sure your bank balance will be sufficient next week, then put Date Paid = 3/15 and Due Date = 3/8.

**Reply by DevAnyKR** [reply]
the old name of this function is bills~.cpp, BILLSDEPOSITS_V1, but in 1.7 the menu and dialog are named Scheduled. Therefore, it should be auto-enter the scheduled transaction rather than for billing purposes. It's not calendar app.


I set it to 2050 because I don't know when I'm going to die, I don't know how long I'm going to use MMEX, and I don't know how long I'm going to see Netflix(= scheduled payment transaction) rates. I want to make it 9999, but I gave up because of an error.

> MMEX will only automatically execute if the due date is <= today

In MMEX, a warning message appears when you change the Due Date to the previous date.
And a stop icon is displayed in the list, indicating that it is overdue.

Due Date must be the date on which auto-run ends.
`if (Today <= due date) then running Scheduled transaction`

And the Date Paid date should be the date that will be automatically entered next time.
`if (Date Paid value <= Today) then automatically execute`
and Date Paid change to next Date( +1 month)

> That field is not when you want the automatic transactions to stop, it's when you want the next transaction to be entered.

MMEX's words contrast with your description.

All I want is for 1 scheduled transaction to be registered when run MMEX. Even if the due date is 2124, not all 100 years' worth will be registered. Because the amount can change along the way. Take Netflix as an example, and the fare goes up little by little.  It's not calendar app.

**Reply by n-stein** [reply]
> Therefore, it should be auto-enter the scheduled transaction

Yes that is exactly what this dialog does. You tell it the transaction details and when to enter it, MMEX automatically enters it.

> I set it to 2050 because I don't know when I'm going to die, I don't know how long I'm going to use MMEX, and I don't know how long I'm going to see Netflix(= scheduled payment transaction) rates. I want to make it 9999, but I gave up because of an error.

You're misunderstanding the way this function works. You don't need to tell it when you want it to stop -- MMEX assumes you want this transaction to be scheduled forever unless you tell it otherwise. You give it two pieces of information, separated on the left and right sides of the dialog:

1. On the left side of the dialog is the scheduling part. You specify when you want the transaction to be entered. This is the "Date Due". It is the date when you want MMEX to automatically enter the transaction. On that date, MMEX will auto enter the transaction. If you set it in the future, MMEX will wait until that date to enter the transaction. You don't want to put 9999-12-31 here, your transaction will never execute because every time MMEX opens it will check "Is it 9999-12-31 yet? No? Ok, guess I don't need to do anything." The date you put here should be the first date you want MMEX to start entering transactions. If you put a date <= today, MMEX will enter the transaction the next time you start the program.

Then you set recurrence. By default MMEX will assume you want to continue entering the transaction at the specified interval until the end of time unless you select Repeats "None" or enter a value in Payments Left (after which MMEX will stop auto entering the transaction). When you open MMEX it will check if the Date Due <= Today and if so it will enter the transaction, then increment the Date Due by your "Repeats" selection. 

For example, today is 2024-03-08. If I schedule a transaction with Date Due 2024-03-09 with Repeats = Monthly then tomorrow on 3/9 when I open MMEX it will automatically enter the transaction and set the Date Due for 2024-04-09. Then on 2024-04-09 when I open MMEX it will automatically enter the transaction again and set the due date to 2024-05-09, etc. it will do this forever unless I set a limit on Payments Left, or set Repeats to "None", or just turn off auto execute. The key here is **_Date Due is when the transaction will be entered next_**. 

Note you can enter a Date Due in the past and MMEX will enter as many transactions as necessary (incrementing the Date Due each time) until the Date Due is in the future. For example, if you enter a Date Due of 2023-03-01 with Repeats "Monthly" MMEX will enter 13 transactions for due dates 2023-03-01, 2023-04-01, 2023-05-01 ... 2024-02-01, and 2024-03-01. At that point it will stop because the next due date is on 2024-04-01 which is in the future.

2. On the right side of the dialog is the transaction details. This is what will be recorded in CHECKINGACCOUNT_V1 on the date due. That's all it is. Here the Date Paid is what will go in the TRANSDATE field. It has no direct connection to the Date Due, you can set it for the past, the future, whatever. When the transaction is auto entered the Date Paid is also auto incremented by the Repeats selection.

So back to the example, if I schedule a transaction with Date Due = 2024-03-09 with Repeats = Monthly and Date Paid = 2024-03-10 then tomorrow on 3/9 when I open MMEX it will automatically enter the transaction with TRANSDATE = 2024-03-10 and set Date Due = 2024-04-09 and Date Paid = 2024-04-10 for the next scheduled transaction. Then on 2024-04-09 when I open MMEX it will automatically enter the transaction with TRANSDATE = 2024-04-10 and set the Date Due to 2024-05-09 and Date Paid to 2024-05-10, etc. The key here is **_Date Paid is what you want in the TRANSDATE field when the transaction is entered_**.

> In MMEX, a warning message appears when you change the Due Date to the previous date.
> And a stop icon is displayed in the list, indicating that it is overdue.

Yes, this makes sense. The Date Due < Today so the transaction is "overdue" -- it should have been entered by now, and would be if auto execute was turned on.

> Due Date must be the date on which auto-run ends.

No, see above.

> And the Date Paid date should be the date that will be automatically entered next time.

Yes, assuming you mean it is the TRANSDATE on the next transaction, not the date the next transaction is entered.

> and Date Paid change to next Date( +1 month)

Yes, and Date Due is also changed by +1 month

> All I want is for 1 scheduled transaction to be registered when run MMEX. 
> Take Netflix as an example, and the fare goes up little by little.

Simple. Create a scheduled transaction. Set both Date Due and Date Paid to the next date you pay Netflix. Set Repeats to Monthly. Turn on auto execute. Next month on that date MMEX will auto enter the transaction.

Sorry if that was long-winded.

**Reply by DevAnyKR** [reply]
Thank you for your detailed explanation 👍👍👍

The reason why I keep asking questions is, 
If set it to 2023-03-01 as above(your example), 13 should be created,
If set it to 2050-12-31, it shouldn't be created automatically, because today is 2024.

But I made dozens or hundreds as attached above, so it's easy to misunderstand if this is a bug or if the due date is an end date.
 (Payments Left value is empty)

**Reply by n-stein** [reply]
> But I made dozens or hundreds as attached above, so it's easy to misunderstand if this is a bug or if the due date is an end date.

Definitely a bug. If the Date Due is too far in the future (+4000 years or so) our calculation result is wrapping around and MMEX is treating it as a past date instead of a future date, hence it auto enters the transaction when it shouldn't.

At least it shouldn't really impact program operation since the Date Due is unlikely to be set that far in the future. But as you noted, we should fix it to prevent "runaway transactions" if a user misconfigures their scheduled transaction.

**Reply by DevAnyKR** [reply]
> You give it two pieces of information, separated on the left and right sides of the dialog:

I understood~😮
left recurring setting and right transactions to register are separate settings.
Only the date is automatically updated.

**Reply by n-stein** [reply]
Let us know if you have any other questions, we appreciate all the bugs you are identifying.

**Reply by DevAnyKR** [reply]
Also, the reason for the misunderstanding is the word.

'due date' word is interpreted as end date(deadline). (Or I thought it was the end date for scheduled work.)
MMEX's real meaning is start base date for schedule

'Date Paid' interprets as the actual transaction date.

I think accounting terms may differ from country to country.

**Comment by DevAnyKR** [comment]
![image](https://github.com/moneymanagerex/moneymanagerex/assets/110871727/e8aca937-1322-4753-95af-af6f8bc38724)

Most booking deals seem to be working fine, but there's still one oddity.
No matter how much you reset the date, future transactions will be registered in advance.
Today is March, and April reservations have been registered.

![image](https://github.com/moneymanagerex/moneymanagerex/assets/110871727/a8a9eb1f-6229-435f-9acb-8667bfc9381a)

I think MMEX recognized the transaction registered before. It automatically moves +1 month.
Even so, future transactions should not be registered.

**Reply by DevAnyKR** [reply]
![image](https://github.com/moneymanagerex/moneymanagerex/assets/110871727/4ea5e89f-d3a9-4149-9d91-10e2a7793ee6)

The date is gradually heading to Andromeda.

Offset calculation is dangerous.

**Comment by trm217** [comment]
@n-stein  I am confused. What is the reason, if even reopening doesn't trigger the automated booking?
![CleanShot 2024-05-27 at 07 31 25@2x](https://github.com/moneymanagerex/moneymanagerex/assets/30313631/6d29df23-b5d4-4fff-a0fb-0b3dd134b639)
Based on the screenshot: the 1st three bookings are set to automatic and should have been executed.
The other two: Why am I not prompted? 😅

**Reply by trm217** [reply]
I just noticed, there also seems to be a bug, that planned transactions seem to duplicate entries in the budget 🤨

**Reply by trm217** [reply]
From how I understood this thread: https://github.com/moneymanagerex/moneymanagerex/issues/5614
The transactions is only triggered after the due-date (Fälligkeitsdatum).
What I don't understand however, even when set to manual confirmation, on the payments day, there is no prompt to confirm the payment 😔
I would really appreciate it, if the documentation for this feature would be enhanced with examples.
